Update RMC generation tools

michaelmackenzie opened the pull request:
Add tools for RMC dataset generation using the modern resampling style. This includes a filter that selects photon conversions within a given region and a generator that resamples these conversions with the G4 Bethe-Heitler spectrum written into a Mu2e Utility.

I've also updated the `GenFilter` filter to additionally filter on energy/momentum, to filter low energy conversions out when resampling photon conversions or generating internal/virtual photon conversions. This will require a change to the Production fcl to keep the behavior the same, as now the filter has multiple purposes so I had to add a flag for filtering the maximum radius in a B-field. I can make a parallel PR for this as well as adding fcl for RMC dataset generation shortly.

FNALbuild commented:
Hi @michaelmackenzie,
You have proposed changes to files in these packages:
- EventGenerator
- CommonMC

which require these tests: build.

@Mu2e/write, @Mu2e/fnalbuild-users have access to CI actions on [main](/Mu2e/Offline/tree/main).





:hourglass: The following tests have been triggered for 8fed6898d692c2f6e90f23fb142fd93236bb84b4: build  (Build queue has 2 jobs)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:sunny: The build tests passed at 8fed6898d692c2f6e90f23fb142fd93236bb84b4.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 8fed6898d692c2f6e90f23fb142fd93236bb84b4 at d8f587081c15e96eb65958ce1bfcfce175e85a99 |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/scons.log). Build time: 04 min 17 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 5 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 304 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 8fed6898d692c2f6e90f23fb142fd93236bb84b4 after being merged into the base branch at d8f587081c15e96eb65958ce1bfcfce175e85a99.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2308/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

brownd1978 commented on EventGenerator/src/GammaConversion_module.cc:
```diff
@@ -0,0 +1,135 @@
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
```
How is this input SimParticle collection created? Is it from standard G4 processes? or do you assume a dedicated module for that (like RMC Generator)? Some description of the requirements on that collection would be helpful. For instance, how do you insure that the output doesn't double-count the conversion?

michaelmackenzie replied:
This collection is created by selecting events with photon conversions using the StoppedParticlesFinder module and saving these conversion events. The GammaConversionPoints filter also only accepts photon conversions, specifically requiring them to be from the process of interest and within a given detector region (r in [0, 500] mm and z in [5000, 8000] mm for the main background estimates). I can add some description in the module that describes the expected input collection. From what I've seen the StoppedParticlesFinder output collection does not include the e+/e- pair from the photon conversion, but I'll try to look closer to ensure this is the case.

brownd1978 commented on EventGenerator/src/GammaConversion_module.cc:
```diff
@@ -0,0 +1,135 @@
+        std::cout << "SimParticle: pdg = " << inpart.pdgId() << " createCode = " << inpart.creationCode()
+                  << " stopCode = " << inpart.stoppingCode() << std::endl;
+      }
+      throw cet::exception("BADINPUT")
```
Why do you require the input collection always has a converted photon? Unless you filter the path leading to this module it seems impossible to guarantee. But maybe that's your intended workflow?

michaelmackenzie replied:
I'll add a comment here, but the current workflow (that I'll make a PR for soon) produces output sim files that require a photon conversion in the event. So all input events should have a photon conversion. If an event doesn't we could pass an empty event as is done in the muon daughter resampling (`SingleProcessGenerator`, though `FlatMuonDaughterGenerator` does throw an error in this case), but this is likely the sign of a problem as we can't resample non-photon conversion events with this generator.

brownd1978 reviewed (commented):
This is a nice addition. I have some general questions about how the photon conversion resampler in particular is intended to work.

michaelmackenzie commented on EventGenerator/src/GenFilter_module.cc:
```diff
@@ -75,42 +109,70 @@ namespace mu2e {
+          GeomHandle<BFieldManager> bfmgr;
+          mu2e::GeomHandle<mu2e::Tracker> tracker;
+          auto tracker_origin = det->toMu2e(tracker->origin());
+          //XYZVectorF pos3Vec = XYZVectorF(aParticle.position().x(),aParticle.position().y(),aParticle.position().z());
```
This section was written by Sophie, but I believe this line can be removed without issue.

michaelmackenzie commented on EventGenerator/src/GenFilter_module.cc:
```diff
@@ -75,42 +109,70 @@ namespace mu2e {
+          KinKal::LoopHelix lh(pos0, mom0, charge, bnom);
+          // calculate rmax and add maxr to siminfo
+          _maxr = sqrt(lh.cx()*lh.cx()+lh.cy()*lh.cy())+fabs(lh.rad());
+          pass_maxr_cut |= _maxr < maxr_max_ && _maxr > maxr_min_;
```
This was written with IPA DIOs in mind I believe, filtering out DIO events that can't make it to the tracker volume. For the conversions, I only use the energy cuts as of now to veto more symmetric conversions. As we'll generate E_gamma > 85 MeV, there's no reason to be saving tracks with E < 84 MeV since we've omitted a relevant part of the photon spectrum already. For RMC as a general background/pileup this cut likely shouldn't be applied, but it's useful towards targeting the high momentum e+/e- component of the background.

michaelmackenzie replied:
By default, each filter option is off. The only change that I had to propagate to current Production fcl is to add `filterMaxR: true` to turn on this filtering, and I set `isNull` to default to false since each individual filter option defaults to false. Turning off all filtering can still be achieved by setting `isNull : true` as before.

sophiemiddleton replied:
Yes, as long as the general default is "no filtering" since this is in the path for all other production primaries too

michaelmackenzie replied:
Yes, the default is still no filtering so other primary production paths should be unaffected by these changes.

brownd1978 replied:
Having all these cuts in a single module raises the possibility they might get used in ways that cause physics biases. For now we can deal with this in production, but before this filter expands again lets think through the design.

michaelmackenzie commented on EventGenerator/src/GenFilter_module.cc:
```diff
@@ -75,42 +109,70 @@ namespace mu2e {
+          mu2e::GeomHandle<mu2e::Tracker> tracker;
+          auto tracker_origin = det->toMu2e(tracker->origin());
+          //XYZVectorF pos3Vec = XYZVectorF(aParticle.position().x(),aParticle.position().y(),aParticle.position().z());
+          ROOT::Math::XYZVector bnom(bfmgr->getBField(tracker_origin));
```
`auto bnom = bfmgr->`  returns a `CLHEP::Hep3Vector` due to `tracker_origin` being this type, while `LoopHelix` requires a `XYZVector` reference so this can't be converted when passing (at least this is the complaint I see in compiling).

michaelmackenzie commented:
I just pushed an update adding some more description in comments. I also added histograms in the `GammaConversion` generator to check that the conversion spectra look reasonable. I found that in my current fcl setup, some `eBrem` and `annih` photons were being stored. I updated the fcl to veto daughters of the "stopped" photons which removes these, and also updated the generator to ignore these when resampling the conversions. I'll make a PR for the Production fcl later today, I'm just finishing testing the generation stages and cleaning the fcl a bit.

michaelmackenzie commented:
I've added a corresponding [Production PR 367](https://github.com/Mu2e/Production/pull/367) that has fcl for the RMC generation and photon conversion resampling.

sdifalco reviewed (approved):
I have spoken to Michael and the PR is well integrating his previous job with the workflow using SimParticle collection as input. We can still keep at the moment the Mu2eG4/src/GammaConvGenerator_module.cc as backup in case we realize the differential distribution dependence on the material is relevant for the pair generation, but for the moment I endorse this generator in the EventGenerator directory

brownd1978 commented on EventGenerator/src/GenFilter_module.cc:
```diff
@@ -75,42 +109,69 @@ namespace mu2e {
+          // make the loophelix
+          KinKal::LoopHelix lh(pos0, mom0, charge, bnom);
+          // calculate rmax and add maxr to siminfo
+          _maxr = sqrt(lh.cx()*lh.cx()+lh.cy()*lh.cy())+fabs(lh.rad());
```
Please use the LoopHelix function for this quantity: https://github.com/KFTrack/KinKal/blob/348e9450e5aead29cc72976015ab12a32884c23d/Trajectory/LoopHelix.hh#L143

michaelmackenzie replied:
I've updated to use the LoopHelix function call. Thank you for the helpful review!

brownd1978 reviewed (approved):
I sign off, pending addressing the one remaining implementation detail.
